compiler/rustc_trait_selection/src/solve/project_goals.rs

```diff
@@ -122,6 +123,14 @@ impl<'tcx> assembly::GoalKind<'tcx> for ProjectionPredicate<'tcx> {
                     )?;
                     ecx.eq(goal.param_env, goal.predicate.term, assumption_projection_pred.term)
                         .expect("expected goal term to be fully unconstrained");
+
+                    // Add GAT where clauses from the trait's definition
```

Comment:
I found it easier to check the GAT WC here in `probe_and_match_goal_against_assumption` rather than *just* doing it for object candidates. This also makes sure we check them correctly for alias bounds, and for most other built-in bounds which end up flowing through this method.

Let me know what you think about this.

Reply:
hm... I don't feel too happy about this but don't really know. It feels potentially unnecessary and might cause issues. But I guess we cna always move it out and check it in less places.

Reply:
I mean, especially moving towards coinduction, I'm not certain if I understand how this isn't going to be necessary. Builtin candidates should be checking their GAT WCs for them to be sound, right?